server/src/main/java/io/druid/server/SegmentManager.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Ordering;
+import com.google.inject.Inject;
+import com.metamx.emitter.EmittingLogger;
+import io.druid.collections.CountingMap;
+import io.druid.segment.ReferenceCountingSegment;
+import io.druid.segment.Segment;
+import io.druid.segment.loading.SegmentLoader;
+import io.druid.segment.loading.SegmentLoadingException;
+import io.druid.timeline.DataSegment;
+import io.druid.timeline.VersionedIntervalTimeline;
+import io.druid.timeline.partition.PartitionChunk;
+import io.druid.timeline.partition.PartitionHolder;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SegmentManager
+{
+  private static final EmittingLogger log = new EmittingLogger(SegmentManager.class);
+
+  private final Object lock = new Object();
+  private final SegmentLoader segmentLoader;
+  private final Map<String, VersionedIntervalTimeline<String, ReferenceCountingSegment>> dataSources = new HashMap<>();
```

Comment:
For the issues which are less related to the original issues, I think it would be fine to fix them in a single PR if the changes are truly little and intuitive, but otherwise, it's better to separate to several PRs even though some PRs will be quite small. This is because 

1) Authors and reviewers can focus on the original issues of the PR. This will increase review speed.
2) As you know, most changes require to add tests even though they are simple. This can be a burden for authors which makes the development slow down.
3) I think getting other people's attention will be good because they can review PRs from other points of view.

Besides, even in the world you mentioned, a lock is needed anyway for synchronization for accessing ```ConcurrentHashMap``` and accessing/mutating ```DataSourceState``` taken from ```ConcurrentHashMap``` (please see ```loadSegment()```). For example,

```java
private final ConcurrentHashMap<String, DataSourceState> dataSources;
...
public boolean loadSegment(final DataSegment segment) throws SegmentLoadingException
{
  synchronized (lock) {
    final DataSourceState dataSourceState = dataSources.get(dataSource);
    final PartitionHolder<ReferenceCountingSegment> entry = dataSourceState.findEntry(segment.getInterval(), segment.getVersion);
    ...
    dataSourceState.add(
        segment.getInterval(),
        segment.getVersion(),
        segment.getShardSpec().createChunk(new ReferenceCountingSegment(adapter))
    );
    ...
  }
}
```

For this issue, I prefer to investigate first what the exact requirements are.